File: applications/wcs-client/app/src/models/wms_model/wms/Layer.ts
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003 - 2017 Peter Baumann /
 rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */

///<reference path="../../../common/_common.ts"/>

module wms {
    export class Layer {        
        // the gml of this layer as a part of result from GetCapabilities request
        // NOTE: WMS does not have a request to describe layer as WCS DescribeCoverage request
        public gmlDocument:string;
        public name:string;
        public title:string;
        public abstract:string;

        // EX_GeographicBoundingBox (reprojected native bounding box to EPSG:4326)        
        public coverageExtent:CoverageExtent;
                
        // native CRS (e.g: EPGS:4326)
        public crs:string;        
        // native bounding box (NOTE: it can be not EPSG:4326)
        public minx:Number;
        public miny:Number;
        public maxx:Number;
        public maxy:Number;
        // display layer's footprint on the globe
        public displayFootprint:boolean;
        
        // layer's styles
        public styles:Style[];
        
        public constructor(gmlDocument:string, name:string, title:string, abstract:string, westBoundLongitude:Number, eastBoundLongitude:Number, 
                           southBoundLatitude:Number, northBoundLatitude:Number, crs:string,
                           minx:Number, miny:Number, maxx:Number, maxy:Number) {
            this.gmlDocument = gmlDocument;
            this.name = name;            
            this.title = title;
            this.abstract = abstract;
            this.coverageExtent = new CoverageExtent(name, westBoundLongitude, southBoundLatitude, eastBoundLongitude, northBoundLatitude);
            this.crs = crs;
            this.minx = minx;
            this.miny = miny;
            this.maxx = maxx;
            this.maxy = maxy;             
            // NOTE: not as WCS, all WMS layers can display on the globe, so no need to set it to null and hide the checkbox when coverage is not geo-referenced.           
            this.displayFootprint = true;

            // build styles from gmlDocument of this layer
            this.buildStylesFromGMLDocument();
        }

        // Extract the Style element of layer to an array
        private buildStylesFromGMLDocument() {
            this.styles = [];
            var tmpXML = $.parseXML(this.gmlDocument);
            var totalStyles = $(tmpXML).find("Style").length;
            for (var i = 0; i < totalStyles; i++) {
                var styleXML = $(tmpXML).find("Style").eq(i);
                var name = styleXML.find("Name").text();
                var abstract = styleXML.find("Abstract").text();
                            
                // parse abstract to know it is rasql transform or wcps fragment query and get the query inside abstract also
                var queryType = 0;
                var query = "";
                var tmp = "";
                if (abstract.indexOf("Rasql transform fragment: ") == -1) {
                    // wcps query fragment
                    queryType = 0;                    
                    tmp = "WCPS query fragment: ";
                } else {
                    // rasql query fragment                    
                    queryType = 1;
                    tmp = "Rasql transform fragment: ";                    
                } 

                // the query (rasql/wcps) for the style
                query = abstract.substring(abstract.indexOf(tmp) + tmp.length, abstract.length);
                var styleAbstract = abstract.substring(0, abstract.indexOf(tmp) - 2).trim();

                this.styles.push(new Style(name, styleAbstract, queryType, query));
            }
        }
    }
}

File: applications/petascope/petascope_core/src/main/java/org/rasdaman/domain/wms/Layer.java
/*
  *  This file is part of rasdaman community.
  * 
  *  Rasdaman community is free software: you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
  *  the Free Software Foundation, either version 3 of the License, or
  *  (at your option) any later version.
  * 
  *  Rasdaman community is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
  *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  *  See the GNU  General Public License for more details.
  * 
  *  You should have received a copy of the GNU  General Public License
  *  along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
  * 
  *  Copyright 2003 - 2014 Peter Baumann / rasdaman GmbH.
  * 
  *  For more information please see <http://www.rasdaman.org>
  *  or contact Peter Baumann via <baumann@rasdaman.com>.
 */
package org.rasdaman.domain.wms;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.Lob;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.OrderColumn;
import javax.persistence.Table;

/**
 *
 * Class representation of a WMS 1.3 layer. NOTE: A layer in WMS is the same
 * model as coverage metadata in WCS. It could have multiple dimensions but the
 * result is always 2D by using the overlay to combine all the slices to 1
 * result.
 *
 * @author <a href="mailto:bphamhuu@jacobs-university.net">Bang Pham Huu</a>
 */
@Entity
@Table(name = Layer.TABLE_NAME)
public class Layer implements Serializable {

    public static final String TABLE_PREFIX = "wms13_";
    public static final String TABLE_NAME = TABLE_PREFIX + "_layer";
    public static final String COLUMN_ID = TABLE_NAME + "_id";
    
    // For Hibernate queries HQL, Criteria
    public static final String LAYER_CLASS_NAME = "Layer";
    public static final String LAYER_NAME_PROPERTY = "name";

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = COLUMN_ID)
    private long id;

    public Layer() {

    }

    // NOTE: each layer can have a parent layer and a parent layer can have multiple child layers (not used now).
    @Column(name = "parent_id")
    private int parent_id;

    // Table 7 — Inheritance of Layer properties
    // One, mandatory
    @Column(name = "name")
    private String name;

    // One, mandatory
    @Column(name = "title")
    private String title;

    // Zero or One, mandatory
    @Column(name = "layer_abstract")
    @Lob
    // NOTE: As this could be long text, so varchar(255) is not enough
    private String layerAbstract;

    // One, but its properties are optional values (use default)
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = LayerAttribute.COLUMN_ID)
    private LayerAttribute layerAttribute;

    // Zero or One, optional
    @ElementCollection(fetch = FetchType.EAGER)
    @OrderColumn
    private List<String> keywordList = new ArrayList<>();

    // Zero Or Many, mandatory
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<Style> styles = new ArrayList<>();

    // One or Many, mandatory
    // List of supported CRS for the layers (e.g :<CRS>EPSG:4326</CRS> <CRS>CRS:84</CRS>)
    // so it will have 2 bounding boxes for each CRS
    @ElementCollection(fetch = FetchType.EAGER)
    @OrderColumn
    private List<String> crss = new ArrayList<>();

    // One, mandatory
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = EXGeographicBoundingBox.COLUMN_ID)
    private EXGeographicBoundingBox exGeographicBoundingBox;

    // One or Many, mandatory
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<BoundingBox> boundingBoxes = new ArrayList<>();

    // Zero Or Many, mandatory
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<Dimension> dimensions = new ArrayList<>();

    // Zero Or One, optional
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = Attribution.COLUMN_ID)
    private Attribution attribution;

    // Zero Or Many, optional
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<AuthorityURL> authorityURLs = new ArrayList<>();

    // Zero Or Many, optional
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<Identifier> identifiers = new ArrayList<>();

    // Zero Or Many, optional
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = Layer.COLUMN_ID)
    @OrderColumn
    private List<MetadataURL> metadataURLs = new ArrayList<>();

    // Zero Or One, optional
    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = DataURL.COLUMN_ID)
    private DataURL dataURL;

    // Constructor with mandatory parameters for WMS layer
    public Layer(String name, String title, String layerAbstract, LayerAttribute layerAttribute,
                 List<Style> styles, List<String> crss, 
                 EXGeographicBoundingBox exGeographicBoundingBox, List<BoundingBox> boundingBoxes, List<Dimension> dimensions) {
        this.name = name;
        this.title = title;
        this.layerAbstract = layerAbstract;
        this.layerAttribute = layerAttribute;
        this.styles = styles;
        this.crss = crss;
        this.exGeographicBoundingBox = exGeographicBoundingBox;
        this.boundingBoxes = boundingBoxes;
        this.dimensions = dimensions;
    }

    public int getParent_id() {
        return parent_id;
    }

    public void setParent_id(int parent_id) {
        this.parent_id = parent_id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getLayerAbstract() {
        return layerAbstract;
    }

    public void setLayerAbstract(String layerAbstract) {
        this.layerAbstract = layerAbstract;
    }

    public LayerAttribute getLayerAttribute() {
        return layerAttribute;
    }

    public void setLayerAttribute(LayerAttribute layerAttribute) {
        this.layerAttribute = layerAttribute;
    }

    public List<String> getKeywordList() {
        return keywordList;
    }

    public void setKeywordList(List<String> keywordList) {
        this.keywordList = keywordList;
    }

    public List<Style> getStyles() {
        return styles;
    }

    public void setStyles(List<Style> styles) {
        this.styles = styles;
    }

    public List<String> getCrss() {
        return crss;
    }

    public void setCrss(List<String> crss) {
        this.crss = crss;
    }

    public EXGeographicBoundingBox getExGeographicBoundingBox() {
        return exGeographicBoundingBox;
    }

    public void setExGeographicBoundingBox(EXGeographicBoundingBox exGeographicBoundingBox) {
        this.exGeographicBoundingBox = exGeographicBoundingBox;
    }

    public List<BoundingBox> getBoundingBoxes() {
        return boundingBoxes;
    }

    public void setBoundingBoxes(List<BoundingBox> boundingBoxes) {
        this.boundingBoxes = boundingBoxes;
    }

    public List<Dimension> getDimensions() {
        return dimensions;
    }

    public void setDimensions(List<Dimension> dimensions) {
        this.dimensions = dimensions;
    }

    public Attribution getAttribution() {
        return attribution;
    }

    public void setAttribution(Attribution attribution) {
        this.attribution = attribution;
    }

    public List<AuthorityURL> getAuthorityURLs() {
        return authorityURLs;
    }

    public void setAuthorityURLs(List<AuthorityURL> authorityURLs) {
        this.authorityURLs = authorityURLs;
    }

    public List<Identifier> getIdentifiers() {
        return identifiers;
    }

    public void setIdentifiers(List<Identifier> identifiers) {
        this.identifiers = identifiers;
    }

    public List<MetadataURL> getMetadataURLs() {
        return metadataURLs;
    }

    public void setMetadataURLs(List<MetadataURL> metadataURLs) {
        this.metadataURLs = metadataURLs;
    }

    public DataURL getDataURL() {
        return dataURL;
    }

    public void setDataURL(DataURL dataURL) {
        this.dataURL = dataURL;
    }

    // Ultility methods
    /**
     * Check if style already existed in list styles of a layer
     *
     * @param styleName
     * @return
     */
    public Style getStyle(String styleName) {
        for (Style style : this.styles) {
            if (style.getName().equals(styleName)) {
                return style;
            }
        }

        return null;
    }
}
